seaborn/_core/data.py

```diff
@@ -285,35 +284,32 @@ def convert_dataframe_to_pandas(data: object) -> pd.DataFrame:
     if isinstance(data, pd.DataFrame):
         return data
 
-    if not hasattr(pd.api, "interchange"):
-        msg = (
-            "Support for non-pandas DataFrame objects requires a version of pandas "
-            "that implements the DataFrame interchange protocol. Please upgrade "
-            "your pandas version or coerce your data to pandas before passing "
-            "it to seaborn."
-        )
-        raise TypeError(msg)
-
-    if _version_predates(pd, "2.0.2"):
-        msg = (
-            "DataFrame interchange with pandas<2.0.2 has some known issues. "
-            f"You are using pandas {pd.__version__}. "
-            "Continuing, but it is recommended to carefully inspect the results and to "
-            "consider upgrading."
-        )
-        warnings.warn(msg, stacklevel=2)
-
-    try:
-        # This is going to convert all columns in the input dataframe, even though
-        # we may only need one or two of them. It would be more efficient to select
-        # the columns that are going to be used in the plot prior to interchange.
-        # Solving that in general is a hard problem, especially with the objects
-        # interface where variables passed in Plot() may only be referenced later
-        # in Plot.add(). But noting here in case this seems to be a bottleneck.
-        return pd.api.interchange.from_dataframe(data)
-    except Exception as err:
-        msg = (
-            "Encountered an exception when converting data source "
-            "to a pandas DataFrame. See traceback above for details."
-        )
-        raise RuntimeError(msg) from err
+    if hasattr(data, '__arrow_c_stream__'):
+        try:
+            import pyarrow
+        except ImportError as err:
+            msg = "PyArrow is required for non-pandas Dataframe support."
```

Comment:
Is this generally a dependency of non-pandas dataframe libraries now? Or could this change introduce a regression for e.g. polars users who are currently leveraging the dataframe interchange protocol?

Reply:
Thanks for your review!

Polars doesn't depend on PyArrow, but `polars.DataFrame.to_pandas` always requires PyArrow. So, in practice, anyone working with both dataframe libraries may well already have PyArrow already installed

To avoid requiring PyArrow for the cases when it's not necessary, one way could be to do something like:
- try using the interchange protocol
- if it raises, then fall back to the PyCapsule Interface (which currently requires PyArrow)

This has the upside of not requiring PyArrow in some cases, but the downside of hiding issues where the interchange protocol silently produces invalid results

It may be possible to do this PyCapsule Interface conversion in the future without PyArrow but with something lighter instead, like arro3 by @kylebarron (who I'm ccing in case he has comments too)

What would be your preference?

Reply:
Some polars users may not have pyarrow installed. If seaborn needs to get pandas data, the only production-ready way to do `Arrow -> pandas` that I know of is using pyarrow. 

As Marco mentions I'm working on [arro3](https://github.com/kylebarron/arro3), which is a minimal library for Arrow in Python, but Pandas interop is not a primary concern, and it's not production-ready today.

Reply:
FWIW pandas 3.x is going to strongly incentivize users to install PyArrow, although it stops short of outright requiring it. In theory, the only people that shouldn't have PyArrow installed are those that operate in space/resource constrained environments, probably in headless environments like AWS Lambda where seaborn won't be used

Of course up to you how much you want to support non-PyArrow configurations, but the dataframe interchange protocol is relatively buggy and gets very little support, so you may find it easier altogether to force users towards PyArrow

Reply:
cuDF have said that they will deprecate the interchange format: https://github.com/rapidsai/cudf/issues/17282

Plotly have stopped using it, so Seaborn is the only library left using it

At this point, I think there's a greater risk in keeping it - I don't want to force anything here of course, just making sure you're aware

Reply:
Sure, happy to drop it. I guess just a question of whether it's better to adopt this new thing or just remove the attempt at cross-dataframe compatibility altogether and leave that up to users.

Reply:
As a user, I find it convenient to be able to pass a dataframe to Seaborn and have it just work, but I can see merits to just telling users to do it themselves too

Reply:
I mean yeah I was on board with the promise of it "just working" but if the infra isn't there yet or there are lots of edge cases it's easier _for me_ to push dealing with them out to the user :)

Reply:
Would you be opposed to taking on the PyArrow dependency? I think if you use that, your implementation becomes much simpler.

The downside to requiring PyArrow if it is not already installed is:

  - PyArrow does not currently distribute musl-c wheels
  - PyArrow requires ~40MB download from pypi and ~130MB on disk

Reply:
Could you please clarify what you mean by "if the infra isn't there yet"?

The Arrow C Interface already has quite widespread adoption and I'm not aware of edge cases in its implementations. @WillAyd wrote about switching his Pantab project over to it in [Leveraging the Arrow C Data Interface](https://willayd.com/leveraging-the-arrow-c-data-interface.html#arrow-c-data-interface-to-the-rescue), and noted

> Almost immediately my issues went away [...] I felt more confident in the implementation and had to deal with less memory corruption / crashes than before. And, perhaps most importantly, I saved a lot of time.

That was nearly a year ago, and given that he's now suggesting it here in Plotly, I'd say that his experience has stayed just as positive

---

Regarding PyArrow dependency, I'll also note that `polars.DataFrame.to_pandas` also requires PyArrow, so any Polars user (such as myself) would already have needed PyArrow installed if they were converting to pandas via the Polars official method

Reply:
Basically my threshold is "do I need to think about it _at all_". I'm just not interested in the minutia of competing Python dataframe libraries or the various attempts to make them work better together. The previous approach was sold as a simple protocol that always works, but it turns out that wasn't the case. Maybe this new way is better, the problem is I have no real way to say for sure without spending a lot of time learning about something that doesn't interest me.

Reply:
Shall I close and leave you to remove cross-dataframe compatibility altogether?

Reply:
😄 that's understandable

I'm aware that you said that using Narwhals was a complete non-starter, but just to showcase that as a possibility:
```python
import narwhals.stable.v1 as nw
from narwhals.stable.v1.typing import IntoDataFrame
import polars as pl
import pandas as pd

def convert_dataframe_to_pandas(data: IntoDataFrame) -> pd.DataFrame:
    return nw.from_native(data).to_pandas()
```
and then leave it up to Narwhals to convert to pandas in the best way for each input library

Altair, Plotly, and Vegafusion are using it as required dependency now, and Bokeh have a PR in progress to do the same